### PR TITLE
[Feature] fix uniform filled in metal

### PR DESCRIPTION
**Pull request by Mee-gu**
fix uniform filled in metal

### CODE REVIEW

File: cocos/renderer/backend/ProgramState.cpp [hunk 1]
```diff
@@ -200,18 +224,72 @@ void ProgramState::setUniform(const backend::UniformLocation& uniformLocation, c
...
         return;
-    
+#ifdef CC_USE_METAL
+    auto uniformInfo = _vertexUniformInfos[location].uniformInfo;
+    if(uniformInfo.needConvert)
+    {
+        convertUniformData(uniformInfo, data, size, _vertexUniformInfos[location].data);
+        return;
+    }
+#endif
```
Comment (minggo) [inline cocos/renderer/backend/ProgramState.cpp:330]: Please add comment about it, why need to do special work for metal.

**Comment by Mee-gu** [comment]
update review

File: cocos/renderer/backend/ProgramState.cpp [hunk 2]
```diff
@@ -200,18 +244,102 @@ void ProgramState::setUniform(const backend::UniformLocation& uniformLocation, c
...
 }
 
 void ProgramState::setFragmentUniform(int location, const void* data, uint32_t size)
 {
     if(location < 0)
         return;
+   
+//float3 etc in Metal has both sizeof and alignment same as float4, need convert to correct laytout
+#ifdef CC_USE_METAL
+    auto uniformInfo = _fragmentUniformInfos[location].uniformInfo;
```
Comment (minggo) [inline cocos/renderer/backend/ProgramState.cpp]: use reference is better

File: cocos/renderer/backend/Types.h [hunk 3]
```diff
@@ -218,13 +218,22 @@ enum class Winding: uint32_t
     COUNTER_CLOCK_WISE
 };
 
+enum class BasicType: uint32_t
```
Comment (minggo) [inline cocos/renderer/backend/Types.h]: Please use more meaningful name

Comment (minggo) [reply cocos/renderer/backend/Types.h]: It is used in uniform data type

**Comment by Mee-gu** [comment]
update review

**Comment by Mee-gu** [comment]
update review

**Review by minggo — APPROVED** [review]

**Comment by minggo** [comment]
please fix ci error

**Comment by Mee-gu** [comment]
close and reopen it to enable CI test again

**Comment by Mee-gu** [comment]
close and reopen it to enable CI test again

**Pull request merged**